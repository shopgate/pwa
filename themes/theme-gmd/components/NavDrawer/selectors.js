/**
 * Selects the menu state from the store.
 * @param {Object} state The global state.
 * @return {Object}
 */
const getMenuState = state => state.menu;

/**
 * Selects a menu by id.
 * @param {Object} state The global state.
 * @param {string} id The menu id.
 * @return {Object}
 */
export const getMenuById = (state, id) => (
  getMenuState(state).menusById[id] || {}
);
