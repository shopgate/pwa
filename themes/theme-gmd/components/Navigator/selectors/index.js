import { createSelector } from 'reselect';
import {
  getHistoryPathname,
  getQueryParamsAsString,
} from '@shopgate/pwa-common/selectors/history';
import {
  getCartProductCount,
  getProductPendingCount,
} from '@shopgate/pwa-common-commerce/cart/selectors';

/**
 * Selects the navigator state.
 * @param {Object} state The global state.
 * @return {Object}
 */
export const getNavigatorState = state => state.navigator;

/**
 * Selects the current background color from navigator state.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getBackgroundColor = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.backgroundColor
);

/**
 * Selects the current text color from navigator state.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getTextColor = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.textColor
);

/**
 * Selects the current state of shadow of the navigator's icons.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isIconShadowShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showIconShadow
);

/**
 * Selects the current state from navigator search.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getSearchPhrase = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.searchPhrase
);

/**
 * Selects the current state from navigator search.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isSearchShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showSearch
);

/**
 * Selects the current state from navigator progress bar.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isProgressBarShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showProgressBar
);

/**
 * Selects the current state from navigator title.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isTitleShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showTitle
);

/**
 * Selects the current state state from navigator disabled status.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isEnabled = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.enabled
);

/**
 * Selects the current state state from navigator disabled status.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isLoadingBarShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showLoadingBar
);

/**
 * Selects the current state of the navigation drawer.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isNavDrawerActive = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.navDrawerActive
);

/**
 * Selects the current state of the navigation drawer.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isNavSearchFieldActive = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.searchActive
);

/**
 * Checks if the cart button is available.
 * @return {boolean}
 */
export const isCartButtonVisible = createSelector(
  getCartProductCount,
  getProductPendingCount,
  getNavigatorState,
  (count, pendingCount, navigator) => (
    navigator.showCartIcon && (count + pendingCount) > 0
  )
);

/**
 * Selects the navigator state for tracking.
 * @param {Object} state The application state.
 * @return {Object}
 */
export const getTrackingNavigatorState = createSelector(
  getHistoryPathname,
  getQueryParamsAsString,
  (pathname, queryString) => ({
    pathname,
    queryString,
  })
);
