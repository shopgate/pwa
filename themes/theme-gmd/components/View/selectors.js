import { createSelector } from 'reselect';

/**
 * Selects the general UI state.
 * @param {Object} state The global state.
 * @return {Object}
 */
export const getGeneralUI = state => state.ui.general;

/**
 * Selects the title from the UI.
 * @param {Object} state The global state.
 * @return {Object}
 */
export const getTitle = createSelector(
  getGeneralUI,
  general => general.title || ''
);

/**
 * Selects the top status from the UI.
 * @param {Object} state The global state.
 * @return {Object}
 */
export const getTopStatus = createSelector(
  getGeneralUI,
  general => general.isTop
);
