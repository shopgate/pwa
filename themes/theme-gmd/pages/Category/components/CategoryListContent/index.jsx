import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import Portal from '@shopgate/pwa-common/components/Portal';
import {
  CATEGORY_LIST,
  CATEGORY_LIST_AFTER,
  CATEGORY_LIST_BEFORE,
} from '@shopgate/pwa-common-commerce/category/constants/Portals';
import { Section } from '@shopgate/engage/a11y';
import { ResponsiveContainer } from '@shopgate/engage/components';
import CategoryList from 'Components/CategoryList';
import CategoryGrid from 'Components/CategoryGrid';
import connect from './connector';

/**
 * The category list content.
 */
class CategoryListContent extends PureComponent {
  static propTypes = {
    categoryId: PropTypes.string.isRequired,
    categories: PropTypes.arrayOf(PropTypes.shape()),
    categoriesFetching: PropTypes.bool,
    childrenCount: PropTypes.number,
    hasChildren: PropTypes.bool,
    showCategoryImages: PropTypes.bool,
  };

  static defaultProps = {
    categories: null,
    categoriesFetching: false,
    childrenCount: 6,
    hasChildren: false,
    showCategoryImages: true,
  }

  /**
   * @returns {JSX}
   */
  render() {
    const {
      hasChildren, categories, categoryId, childrenCount, categoriesFetching,
    } = this.props;

    return (
      <Fragment>
        <Portal name={CATEGORY_LIST_BEFORE} props={{ categoryId }} />
        <Portal name={CATEGORY_LIST} props={{ categoryId }}>
          {hasChildren && (
            <Section title="category.sections.categories">
              <ResponsiveContainer appAlways breakpoint="<=xs">
                <CategoryList
                  categories={categories}
                  prerender={categoriesFetching ? childrenCount : 0}
                  showImages={this.props.showCategoryImages}
                />
              </ResponsiveContainer>
              <ResponsiveContainer webOnly breakpoint=">xs">
                <CategoryGrid
                  categories={categories}
                  prerender={categoriesFetching ? childrenCount : 0}
                  showImages={this.props.showCategoryImages}
                />
              </ResponsiveContainer>
            </Section>
          )}
        </Portal>
        <Portal name={CATEGORY_LIST_AFTER} props={{ categoryId }} />
      </Fragment>
    );
  }
}

export default connect(CategoryListContent);
