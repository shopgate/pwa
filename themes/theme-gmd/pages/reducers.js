import { combineReducers } from 'redux';
import client from '@shopgate/pwa-common/reducers/client';
import history from '@shopgate/pwa-common/reducers/history';
import url from '@shopgate/pwa-common/reducers/url';
import user from '@shopgate/pwa-common/reducers/user';
import page from '@shopgate/pwa-common/reducers/page';
import view from '@shopgate/pwa-common/reducers/view';
import menu from '@shopgate/pwa-common/reducers/menu';
import modal from '@shopgate/pwa-common/reducers/modal';
import toast from '@shopgate/pwa-common/reducers/toast';
import cart from '@shopgate/pwa-common-commerce/cart/reducers';
import category from '@shopgate/pwa-common-commerce/category/reducers';
import favorites from '@shopgate/pwa-common-commerce/favorites/reducers';
import filter from '@shopgate/pwa-common-commerce/filter/reducers';
import product from '@shopgate/pwa-common-commerce/product/reducers';
import search from '@shopgate/pwa-common-commerce/search/reducers';
import reviews from '@shopgate/pwa-common-commerce/reviews/reducers';
import navigator from 'Components/Navigator/reducer';
import extensions from 'Extensions/reducers';
import general from 'Components/View/reducer';
import categoryPage from 'Pages/Category/reducer';

const reducers = {
  cart,
  category,
  client,
  ...extensions && { extensions: combineReducers(extensions) },
  favorites,
  filter,
  history,
  menu,
  modal,
  navigator,
  page,
  product,
  reviews,
  search,
  toast,
  ui: combineReducers({
    general,
    categoryPage,
  }),
  url,
  user,
  view,
};

export default reducers;
