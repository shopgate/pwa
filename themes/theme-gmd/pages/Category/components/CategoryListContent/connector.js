import { connect } from 'react-redux';
import isEqual from 'lodash/isEqual';
import {
  getCategoryChildren,
  areCategoryChildrenFetching,
  getCategoryChildCount,
  hasCategoryChildren,
  getCategory,
} from '@shopgate/pwa-common-commerce/category/selectors';
import { hasNewServices } from '@shopgate/engage/core/helpers';
import { getShowCategoryImages } from '@shopgate/engage/core/selectors/shopSettings';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @param {Object} props The component props.
 * @return {Object} The extended component props.
 */
const mapStateToProps = (state, props) => ({
  category: getCategory(state, props),
  categories: getCategoryChildren(state, props),
  categoriesFetching: areCategoryChildrenFetching(state, props),
  hasChildren: hasCategoryChildren(state, props),
  childrenCount: getCategoryChildCount(state, props),
  showCategoryImages: hasNewServices() ? getShowCategoryImages(state, props) : false,
});

/**
 * @param {Object} next The next component props.
 * @param {Object} prev The previous component props.
 * @returns {boolean}
 */
const areStatePropsEqual = (next, prev) => {
  if (prev.hasChildren !== next.hasChildren) {
    return false;
  }

  if (!isEqual(prev.categories, next.categories)) {
    return false;
  }

  if (prev.childrenCount !== next.childrenCount) {
    return false;
  }

  return true;
};

export default connect(mapStateToProps, null, null, { areStatePropsEqual });
