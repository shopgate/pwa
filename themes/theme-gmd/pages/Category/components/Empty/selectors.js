import { createSelector } from 'reselect';
import {
  getCategoryProductCount,
  getCategoryChildCount,
} from '@shopgate/pwa-common-commerce/category/selectors';
import { getProductsResult } from '@shopgate/pwa-common-commerce/product/selectors/product';

/**
 * Determines whether or not a category has any content,
 * (subcategories, child products or products received via request).
 */
export const isCategoryEmpty = createSelector(
  getCategoryProductCount,
  getCategoryChildCount,
  getProductsResult,
  (productCount, childrenCount, { products, totalProductCount }) => {
    if (childrenCount !== 0) {
      return false;
    }

    if (productCount === 0) {
      return true;
    }

    if (totalProductCount !== null && products.length === 0) {
      return true;
    }

    return false;
  }
);
