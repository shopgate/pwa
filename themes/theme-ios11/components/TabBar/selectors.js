import { createSelector } from 'reselect';
import { getCurrentPathname } from '@shopgate/pwa-common/selectors/router';
import { INDEX_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import { CART_PATH } from '@shopgate/pwa-common-commerce/cart/constants';
import { FAVORITES_PATH } from '@shopgate/pwa-common-commerce/favorites/constants';
import { CATEGORY_PATH } from '@shopgate/pwa-common-commerce/category/constants';
import { SEARCH_PATH } from '@shopgate/pwa-common-commerce/search/constants';
import { BROWSE_PATH } from 'Pages/Browse/constants';
import { MORE_PATH } from 'Pages/More/constants';
import { getIsCookieConsentHandled } from '@shopgate/engage/tracking/selectors/cookieConsent';
import {
  TAB_HOME,
  TAB_BROWSE,
  TAB_CART,
  TAB_FAVORITES,
  TAB_MORE,
  TAB_NONE,
} from './constants';

/**
 * Returns a tabBar state.
 * @param {Object} state State.
 * @return {Object}
 */
const getTabBarState = state => state.ui.tabBar;

/**
 * Returns what tab is active, returns TAB_NONE if none is active.
 * @param {Object} state The application state.
 * @returns {string}
 */
export const getActiveTab = createSelector(
  getCurrentPathname,
  (pathname) => {
    if (!pathname) {
      return TAB_NONE;
    }

    switch (true) {
      case pathname === INDEX_PATH:
        return TAB_HOME;
      case (pathname === BROWSE_PATH
        || pathname.startsWith(SEARCH_PATH)
        || pathname.startsWith(CATEGORY_PATH)):
        return TAB_BROWSE;
      case pathname === CART_PATH:
        return TAB_CART;
      case pathname === MORE_PATH:
        return TAB_MORE;
      case pathname === FAVORITES_PATH:
        return TAB_FAVORITES;
      default:
        return TAB_NONE;
    }
  }
);

/**
 * Checks if the tab bar is currently enabled.
 * @return {boolean}
 */
export const isTabBarEnabled = createSelector(
  getTabBarState,
  getIsCookieConsentHandled,
  // Do not show the TabBar when cookie consent is not handled yet. This prevents breaking out
  // of the cookie consent process.
  (state, cookieConsentHandled) => cookieConsentHandled && state.enabled
);

/**
 * Checks if the tab bar is currently visible.
 * @return {boolean}
 */
export const isTabBarVisible = createSelector(
  getTabBarState,
  isTabBarEnabled,
  (state, enabled) => {
    if (!enabled) {
      return false;
    }

    return state.visible;
  }
);
