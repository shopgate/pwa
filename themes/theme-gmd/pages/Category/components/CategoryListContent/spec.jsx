import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { createMockStore } from '@shopgate/pwa-common/store';
import { hasCategoryChildren } from '@shopgate/engage/category/selectors';
import reducers from 'Pages/reducers';
import CategoryListContent from './index';

const store = createMockStore(reducers);

jest.mock('@shopgate/pwa-common-commerce/category/selectors', () => ({
  ...jest.requireActual('@shopgate/pwa-common-commerce/category/selectors'),
  hasCategoryChildren: jest.fn().mockReturnValue(false),
}));
jest.mock('@shopgate/engage/a11y', () => ({
  Section: ({ children }) => children,
}));

jest.mock('@shopgate/engage/components');
jest.mock('@shopgate/engage/category/components', () => ({
  CategoryList: () => null,
}));
jest.mock('Components/CategoryGrid', () => ({
  __esModule: true,
  default: function CategoryGrid() { return null; },
}));

describe('<CategoryListContent />', () => {
  it('should render without CategoryList', () => {
    const wrapper = mount((
      <Provider store={store}>
        <CategoryListContent categoryId="1234" />
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CategoryList')).toHaveLength(0);
  });

  it('should render with CategoryList', () => {
    hasCategoryChildren.mockReturnValueOnce(true);
    const wrapper = mount((
      <Provider store={store}>
        <CategoryListContent categoryId="1234" />
      </Provider>
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('CategoryList')).toHaveLength(1);
  });
});
