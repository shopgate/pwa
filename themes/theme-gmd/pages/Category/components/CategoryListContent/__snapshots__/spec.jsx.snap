// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryListContent /> should render with CategoryList 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CategoryListContent)
    categoryId="1234"
  >
    <CategoryListContent
      categories={null}
      categoriesFetching={false}
      category={null}
      categoryId="1234"
      childrenCount={null}
      dispatch={[Function]}
      hasChildren={true}
      showCategoryImages={false}
    >
      <Portal
        name="category.list.before"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
      <Portal
        name="category.list"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      >
        <Section
          title="category.sections.categories"
        >
          <ResponsiveContainer
            appAlways={true}
            breakpoint="<=xs"
          >
            <CategoryList
              categories={null}
              parentCategory={null}
              prerender={0}
              showImages={false}
            />
          </ResponsiveContainer>
          <ResponsiveContainer
            breakpoint=">xs"
            webOnly={true}
          >
            <CategoryGrid
              categories={null}
              prerender={0}
              showImages={false}
            />
          </ResponsiveContainer>
        </Section>
      </Portal>
      <Portal
        name="category.list.after"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
    </CategoryListContent>
  </Connect(CategoryListContent)>
</Provider>
`;

exports[`<CategoryListContent /> should render without CategoryList 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CategoryListContent)
    categoryId="1234"
  >
    <CategoryListContent
      categories={null}
      categoriesFetching={false}
      category={null}
      categoryId="1234"
      childrenCount={null}
      dispatch={[Function]}
      hasChildren={false}
      showCategoryImages={false}
    >
      <Portal
        name="category.list.before"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
      <Portal
        name="category.list"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
      <Portal
        name="category.list.after"
        props={
          Object {
            "categoryId": "1234",
          }
        }
      />
    </CategoryListContent>
  </Connect(CategoryListContent)>
</Provider>
`;
